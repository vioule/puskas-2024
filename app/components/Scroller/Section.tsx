import { useInView } from "framer-motion";
import { useRef, useEffect, useContext } from "react";
import { AppContext, DataType } from "../DataProvider";

interface SectionProps {
  data: DataType;
}

export default function Section({ data }: SectionProps) {
  const ref = useRef(null);
  const { setData, lightbox, setLightbox } = useContext(AppContext);
  const timer = useRef<number | undefined>(undefined);
  const isInView = useInView(ref, { amount: 0.7, once: false });
  const color = isInView ? "text-white" : "text-gold";
  const cursor = lightbox.preview ? "hover:cursor-pointer" : "";

  const handleOnClick = () => {
    if (lightbox.preview) {
      setLightbox({
        preview: true,
        poster: true,
      });
    }
  };

  useEffect(() => {
    if (!lightbox.poster) {
      if (isInView) {
        timer.current = window.setTimeout(() => {
          setData(data);
          setLightbox({
            preview: true,
            poster: false,
          });
        }, 400);
      } else {
        window.clearTimeout(timer.current);
        setLightbox({
          preview: false,
          poster: false,
        });
      }
    }
    return () => clearTimeout(timer.current);
  }, [isInView, setData, data, setLightbox, lightbox.poster]);
  return (
    <div
      className={`flex justify-center items-center w-full h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] max-w-[800px] transition-[color] duration-1000 ease-linear ${color} ${cursor}`}
      ref={ref}
      onClick={handleOnClick}
      data-testid="section"
    >
      <span>{data.year}</span>
    </div>
  );
}
