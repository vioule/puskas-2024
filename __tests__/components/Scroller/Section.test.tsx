import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Section from "@/app/components/Scroller/Section";
import {
  AppContext,
  DataType,
  LightboxType,
} from "@/app/components/DataProvider";
import { useInView } from "framer-motion";

jest.mock("framer-motion");
const dataInitialState: DataType = {
  year: 2009,
  name: "cristiano ronaldo",
  comp: "UEFA champions league",
};
const setData = jest.fn();
const setLightbox = jest.fn();

describe("<Section />", () => {
  const mockedUseInView = jest
    .mocked(useInView)
    .mockReturnValueOnce(true)
    .mockReturnValueOnce(true)
    .mockReturnValueOnce(true)
    .mockReturnValueOnce(true)
    .mockReturnValue(false);

  test("render Section unchanged", async () => {
    const { container } = render(<Section data={dataInitialState} />);
    expect(container).toMatchSnapshot();
  });

  describe("inView !lightbox.preview !lightbox.poster", () => {
    const lightbox: LightboxType = {
      preview: false,
      poster: false,
    };
    beforeEach(() => {
      render(
        <AppContext.Provider
          value={{
            data: dataInitialState,
            setData,
            lightbox,
            setLightbox,
          }}
        >
          <Section data={dataInitialState} />
        </AppContext.Provider>
      );
    });

    afterEach(() => {
      mockedUseInView.mockClear();
      setData.mockClear();
      setLightbox.mockClear();
    });

    test("Section onClick should not trigger", async () => {
      fireEvent.click(screen.getByTestId("section"));
      expect(setLightbox).not.toHaveBeenCalled();
    });

    test("Section InView correctly send state", async () => {
      await waitFor(() => {
        expect(setLightbox.mock.calls).toEqual([
          [
            {
              preview: true,
              poster: false,
            },
          ],
        ]);
        expect(setData.mock.calls).toEqual([[dataInitialState]]);
      });
    });

    test("should have correct styles", async () => {
      const section = screen.getByTestId("section");
      expect(section).toHaveClass("text-white");
      expect(section).not.toHaveClass("text-gold");
    });
  });

  describe("!inView lightbox.preview !lightbox.poster", () => {
    const lightbox: LightboxType = {
      preview: true,
      poster: false,
    };
    beforeEach(() => {
      render(
        <AppContext.Provider
          value={{
            data: dataInitialState,
            setData,
            lightbox,
            setLightbox,
          }}
        >
          <Section data={dataInitialState} />
        </AppContext.Provider>
      );
    });

    afterEach(() => {
      mockedUseInView.mockClear();
      setData.mockClear();
      setLightbox.mockClear();
    });

    test("Section !inView should correctly send state", async () => {
      expect(setLightbox.mock.calls).toEqual([
        [
          {
            preview: false,
            poster: false,
          },
        ],
      ]);
    });
    test("Section onClick should trigger", async () => {
      fireEvent.click(screen.getByTestId("section"));
      expect(setLightbox.mock.lastCall).toEqual([
        {
          preview: true,
          poster: true,
        },
      ]);
    });
    test("should have correct styles", async () => {
      expect(screen.getByTestId("section")).toHaveClass("hover:cursor-pointer");
    });
  });
  describe("!inView !lightbox.preview lightbox.poster", () => {
    const lightbox: LightboxType = {
      preview: false,
      poster: true,
    };
    beforeEach(() => {
      render(
        <AppContext.Provider
          value={{
            data: dataInitialState,
            setData,
            lightbox,
            setLightbox,
          }}
        >
          <Section data={dataInitialState} />
        </AppContext.Provider>
      );
    });

    afterEach(() => {
      mockedUseInView.mockClear();
      setData.mockClear();
      setLightbox.mockClear();
    });

    test("should not send state", async () => {
      expect(setLightbox).not.toHaveBeenCalled();
      expect(setData).not.toHaveBeenCalled();
    });
    test("should have correct styles", async () => {
      const section = screen.getByTestId("section");
      expect(section).not.toHaveClass("hover:cursor-pointer");
      expect(section).toHaveClass("text-gold");
      expect(section).not.toHaveClass("text-white");
    });
  });
});
