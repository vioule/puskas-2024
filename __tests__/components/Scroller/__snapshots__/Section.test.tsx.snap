// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Section /> render Section unchanged 1`] = `
<div>
  <div
    class="flex justify-center items-center w-full h-screen font-rocksalt text-7xl tracking-[1.5rem] first:mt-[100vh] max-w-[800px] transition-[color] duration-1000 ease-linear text-white "
    data-testid="section"
  >
    <span>
      2009
    </span>
  </div>
</div>
`;
